ImputationPipeline/CreateAggregationSets/CreateAggregationSets.py

```diff
@@ -0,0 +1,235 @@
+                else:
+                    lab_batch_sample_sets_dict[attributes['lab_batch']] = this_aggregation_set
+
+        # Read samples from samples table
+        print(f'Reading {table_name} table...')
+        sample_response = fapi.get_entities(self.workspace_namespace, self.workspace_name, f'{table_name}')
+        if not sample_response.ok:
+            raise RuntimeError(f'ERROR: {sample_response.text}')
+
+        samples = json.loads(sample_response.text)
+        # Writing new sample_set_membership.tsv
+        added_sample_sets_dict = dict()  # dictionary from lab_batch to aggregation sets with added samples
+        control_samples_dict = dict()  # dictionary from lab_batch to sample id of control sample
+        added_samples_dict = dict()  # dictionary from set_id to list of samples to be added to the set
+        with open(f'CreateSampleSets_data/new_{table_name}_set_membership.tsv', 'w') as new_membership_file:
+            # Write header
+            new_membership_file.write(f'membership:{table_name}_set_id\t{table_name}\n')
+            for sample in samples:
+                if 'lab_batch' not in sample['attributes']:
+                    continue
+                sample_name = sample['name']
+                lab_batch = sample['attributes']['lab_batch']
+                is_control_sample = sample['attributes']['is_control_sample']
+                rework = sample['attributes'].get('rework', False)
+                if is_control_sample:
+                    # do we already have a control sample for this lab batch?  that would be bad...
+                    if lab_batch in control_samples_dict:
+                        raise RuntimeError(
+                            f'Multiple control samples for lab_bath {lab_batch}: {sample_name}, {control_samples_dict[lab_batch]}')
+                    # store control sample name in dictionary
+                    control_samples_dict[lab_batch] = sample_name
+                    # we do not create aggregation sets if we only have the control sample.  We will add control samples to newly created aggregation sets later
+                    continue
+                if rework or sample_name not in samples_already_in_aggregation_sets:
+                    # this (non-control) sample needs to be added to an aggregation set.
+                    # Find (or create) aggregation set to add it to.
+                    if lab_batch not in added_sample_sets_dict:
+                        if lab_batch in lab_batch_sample_sets_dict:
+                            previous_aggregation_set = lab_batch_sample_sets_dict[lab_batch]
+                            if previous_aggregation_set.delivered:
+                                # we need to create the next aggregation set for this lab batch
+                                added_sample_sets_dict[lab_batch] = next_aggregation_set(previous_aggregation_set)
```

Comment:
Also here, wouldn't it be easier (and more readable) to just call it in-line and get rid of the two helper methods?